### PR TITLE
fix: prevent sidebar flash on session detail pages

### LINKED ISSUE #1230: web: sidebar flashes 'No active sessions' on /sessions/[id] and re-renders every 5s
## Symptom

When navigating to `/sessions/[id]`, the project sidebar briefly shows **"No active sessions"** before the real session rows pop in a few seconds later. Additionally, the sidebar re-renders every 5 seconds even when nothing has changed.

This does **not** happen on `/` (the dashboard) — only on session detail pages.

## Root cause

`packages/web/src/app/sessions/[id]/page.tsx` is a pure `"use client"` component with no SSR. On mount it seeds:

```ts
const [projects, setProjects] = useState<ProjectInfo[]>([]);
const [sidebarSessions, setSidebarSessions] = useState<DashboardSession[]>([]);
```

So the very first paint is, by construction, an empty sidebar. `packages/web/src/components/ProjectSidebar.tsx:237` then renders `"No active sessions"` because `visibleSessions.length === 0`. The "No active sessions" copy is meant as a *terminal* empty state (the project genuinely has zero work), not as a loading state — using it as a transient placeholder lies to the user on every navigation.

After mount, four separate fetches fire from separate `useEffect`s:
- `/api/projects` (mount)
- `/api/sessions/:id` (mount + 5s poll)
- `/api/sessions` for the sidebar (mount + 5s poll)
- `/api/sessions?project=…&orchestratorOnly=true` (after a 2s `setTimeout` + 5s poll)

Once `/api/sessions` resolves, `setSidebarSessions(body?.sessions ?? [])` runs and the sidebar fills in. That's the flash.

The 5s `setInterval` then re-runs `setSidebarSessions(...)` **unconditionally** with a brand-new array reference, so `ProjectSidebar` re-renders and its `useMemo`s recompute every 5 seconds even when nothing actually changed.

For comparison, `/` (`packages/web/src/app/page.tsx`) is a server component that pre-renders sessions via `getDashboardPageData()` and pipes them through `useSessionEvents()` (which consumes the `MuxProvider` `sessions` topic for live updates). No flash, no polling. The session detail route is the architectural odd one out.

## Why this matters

- **P0** for perceived quality. The sidebar is the primary navigation surface, and it contradicts itself on every session click.
- Layout shift + content shift + semantic shift on every navigation erodes trust.
- The 5s re-render also makes the sidebar feel "alive" in a bad way and burns battery on idle pages.

## Expected behavior

- First paint already shows the populated sidebar (or skeleton rows during loading — never the terminal "No active sessions" copy as a placeholder).
- No re-render when underlying session data is unchanged.
- One data path for sidebar updates, shared with `/`.

## Fix

Two paths exist depending on timing:

### Long-term (already implemented)
The `ashish/feat/multi-dashboard-final` branch already fixes this structurally:
- Deletes `app/sessions/[id]/page.tsx` in favor of a server-rendered `app/projects/[projectId]/sessions/[sessionId]/page.tsx`
- Hosts the sidebar inside a shared `<DashboardShell>` that pipes SSR sessions through `useSessionEvents()` for live updates
- Uses `<UnifiedSidebar>` instead of the ad-hoc per-route `ProjectSidebar` wiring

When that branch merges, this bug auto-resolves.

### Short-term (until that branch lands)
Minimal patch to `packages/web/src/app/sessions/[id]/page.tsx`, contained entirely in code the multi-project branch deletes (zero merge conflict risk):

1. Initialize `sidebarSessions` to `null` (not `[]`) and pass it to `ProjectSidebar` with a separate `loading` prop. Show skeleton rows while loading; reserve `"No active sessions"` for the genuine empty case.
2. Parallelize the initial mount fetches via `Promise.all` so `/api/sessions` doesn't queue behind `/api/projects` + `fetchSession`.
3. In `fetchSidebarSessions`, diff the new array against the previous one by `id + status + activity + lastActivityAt`. Skip `setSidebarSessions` if unchanged. Kills the every-5s re-render and `useMemo` recomputation.

Optional follow-up: drop the 5s sidebar poll entirely and subscribe to `useMux().sessions` instead — same source `Dashboard` already uses via `useSessionEvents`.

## Affected files

- `packages/web/src/app/sessions/[id]/page.tsx` — the route with the bug
- `packages/web/src/components/ProjectSidebar.tsx:237` — where "No active sessions" is rendered
- `packages/web/src/providers/MuxProvider.tsx` — already has the live `sessions` topic the polling could be replaced with
- `packages/web/src/hooks/useSessionEvents.ts` — the hook `/` uses; the session detail route should also use it

### PR DESCRIPTION
**Pull request by ashish921998**
## Summary
- keep session-detail sidebar data nullable until it loads and render sidebar skeleton rows instead of the terminal empty state
- parallelize the initial session/projects/sidebar fetches and stop replacing sidebar session state when the polled payload is unchanged
- add coverage for the new loading path and align the existing terminal-link test with current SessionCard behavior

Closes #1230

## Validation
- pnpm build
- pnpm typecheck
- pnpm lint
- pnpm test
- pnpm --filter @aoagents/ao-web test

### CODE REVIEW

**Review by i-trytoohard — CHANGES_REQUESTED** [review]
## Code Review — PR #1232

**Title:** fix: prevent sidebar flash on session detail pages
**Author:** @ashish921998
**Scope:** 6 files, +563/-39 — entirely web package, focused on `/sessions/[id]`

Closes #1230 — sidebar flashing "No active sessions" on load + redundant re-renders every 5s.

---

### Overall Assessment: **Solid fix, clean execution.** One real issue to address before merge.

The PR nails the three goals from #1230:
1. ✅ Sidebar now shows skeleton rows while loading (null state), reserves "No active sessions" for the genuine empty case
2. ✅ Initial fetches parallelized via `Promise.all` instead of sequential
3. ✅ Shallow-equality checks prevent no-op re-renders from polling
4. ✅ MuxProvider integration for real-time patches between poll cycles — nice bonus

CI all green. Test coverage is thorough (+306 lines of tests for the new loading/error/mux paths).

---

### 🔴 Issue — Mux effect has `sidebarSessions` in its dependency array

**File:** `packages/web/src/app/sessions/[id]/page.tsx` — the mux `useEffect`

```ts
}, [fetchSidebarSessions, mux?.sessions, sidebarSessions]);
```

`sidebarSessions` is state that changes every time data updates. Including it in the mux effect deps means this effect re-fires on every data change, which:
- Re-runs `applyMuxSessionPatches` against the same mux snapshot it already applied
- Potentially calls `fetchSidebarSessions()` again if the ID-set check detects a delta that was already resolved

Since the mux effect only needs to react to **new mux data arriving** (not to sidebar state changes), `sidebarSessions` should be removed from the dependency array. Read `sidebarSessions` via a ref if needed inside the effect body, or read from `cachedSidebarSessions` (module-level variable that's always current).

The initial fetch effect deps (`[fetchProjects, fetchSession, fetchSidebarSessions]`) are fine — those are `useCallback` with `[]` deps, so they're stable.

---

### ⚠️ Warning — `areSidebarSessionsEqual` uses `JSON.stringify` for deep comparison

**File:** `page.tsx` lines ~54-72

The comparison function does `JSON.stringify(project) === JSON.stringify(candidate)` and same for sessions. This works but is:
- **O(n²)** in the worst case for large session lists (though probably fine at AO scale)
- Brittle if property ordering changes on the server

A proper deep-equal or field-level comparison would be more robust. For sessions, you already know the key fields (`id`, `status`, `activity`, `lastActivityAt`) — comparing just those would be more predictable.

**Note:** Cursor Bugbot flagged that `areSidebarSessionsEqual` only compares a few fields and misses PR-level changes. That's actually a valid concern — but for this short-term patch it's acceptable since the 5s poll will eventually catch those changes. Just worth noting for the long-term `UnifiedSidebar` migration.

---

### ⚠️ Warning — Module-level caches survive HMR and navigation

```ts
let cachedProjects: ProjectInfo[] | null = null;
let cachedSidebarSessions: DashboardSession[] | null = null;
```

These are module-level variables. They survive component remounts (good — that's the intent) but also survive HMR and could serve stale data in dev. Not a blocker, but worth a `// TODO: clear on HMR` comment or a dev-mode guard.

---

### 💡 Suggestions (non-blocking)

1. **`applyMuxSessionPatches` validation warning dedup** — `warnedMuxPatchValues` is a module-level Set that grows forever. At AO scale this is fine, but consider capping it or using a dev-only guard.

2. **The 5s poll still runs `fetchSidebarSessions`** — now that mux patches arrive in real-time, the poll is somewhat redundant for sidebar data. The mux effect already triggers a full re-fetch when new session IDs appear. Could skip the sidebar poll and rely solely on mux → but that's a follow-up, not this PR's responsibility.

3. **Test for `areProjectsEqual` / `areSidebarSessionsEqual`** — these are exported as plain functions but have no direct unit tests. They're covered indirectly through the integration tests, but a couple of targeted edge-case tests (empty arrays, reordered items) would be cheap insurance.

---

### ✅ What Looks Good

- **Skeleton UX is well-done** — 3 pulsing rows per project with proper `aria-label="Loading sessions"`. Clean, accessible.
- **Error state with retry button** — nice touch, handles the `/api/sessions` failure case gracefully.
- **`applyMuxSessionPatches` is well-structured** — validates status/activity values, dedupes warnings, short-circuits when nothing changed.
- **Mux integration handles the race condition** — pending patches held in a ref, applied when the REST fetch resolves. The test for this ("applies mux snapshots that arrive before the initial sidebar fetch resolves") is exactly the right scenario to cover.
- **`sidebarSessions` initialized from cache** — `useState(() => cachedSidebarSessions)` avoids the flash on remount. Smart.
- **Component prop interface is clean** — `sidebarLoading`, `sidebarError`, `onRetrySidebar` are straightforward additions to `SessionDetailProps` and `ProjectSidebarProps`.

---

**Verdict: Request changes** on the mux effect dependency array issue. Everything else is solid.

File: packages/web/src/app/sessions/[id]/page.tsx [hunk 1]
```diff
@@ -183,36 +303,81 @@ export default function SessionPage() {
...
+      return;
+    }
+
+    for (const muxId of muxIds) {
+      if (!cachedIds.has(muxId)) {
+        void fetchSidebarSessions();
+        return;
+      }
+    }
+  }, [fetchSidebarSessions, mux?.sessions, sidebarSessions]);
```
Comment (i-trytoohard) [inline packages/web/src/app/sessions/[id]/page.tsx]: 🔴  in the dependency array causes this effect to re-fire on every data change, not just when mux data arrives. Since you already read  inside , and the function short-circuits when nothing changed, this is mostly harmless — but it still triggers the ID-set comparison and potentially calls  redundantly.

Use a ref for  inside this effect, or read from  (module-level). Remove  from the dep array.

**Review by i-trytoohard — DISMISSED** [review]

File: packages/web/src/app/sessions/[id]/page.tsx (hunk 1 above)
Comment (ashish921998) [reply packages/web/src/app/sessions/[id]/page.tsx]: Fixed in 41797001 — the mux effect now reads from the module-level `cachedSidebarSessions` and `sidebarSessions` is removed from the dep array, so it only re-fires when new mux data arrives.

File: packages/web/src/app/sessions/[id]/page.tsx [hunk 2]
```diff
@@ -52,6 +52,25 @@ interface ProjectSessionsBody {
...
+  return previous.every((session, index) => {
+    const candidate = next[index];
+    return (
+      session.id === candidate?.id &&
+      session.status === candidate.status &&
+      session.activity === candidate.activity &&
+      session.lastActivityAt === candidate.lastActivityAt
+    );
+  });
+}
```
[the remark this answers is not in the document]
Comment (ashish921998) [reply packages/web/src/app/sessions/[id]/page.tsx:94]: Resolved in a5e4b13a — `areSidebarSessionsEqual` now uses a full `JSON.stringify` deep compare on each session, so PR-level changes (state, mergeability, CI status, reviewDecision, unresolvedThreads, branch, title) are all picked up.

File: packages/web/src/app/sessions/[id]/page.tsx [hunk 3]
```diff
@@ -193,26 +263,65 @@ export default function SessionPage() {
...
-    // Delay zone counts so the heavy /api/sessions call doesn't contend with session load
-    const t = setTimeout(fetchProjectSessions, 2000);
-    return () => clearTimeout(t);
-  }, [fetchSession, fetchProjectSessions, fetchSidebarSessions]);
+    void Promise.all([
+      fetchProjects(),
+      fetchSession(),
+      sidebarSessions === null ? fetchSidebarSessions() : Promise.resolve(),
+    ]);
+  }, [fetchProjects, fetchSession, fetchSidebarSessions, sidebarSessions]);
```
[the remark this answers is not in the document]
Comment (ashish921998) [reply packages/web/src/app/sessions/[id]/page.tsx]: Resolved in a5e4b13a — the initial-fetch effect no longer has `sidebarSessions` in its dep array. It's now `[fetchProjects, fetchSession, fetchSidebarSessions]`, all stable `useCallback`s, so it runs once on mount.

File: packages/web/src/app/sessions/[id]/page.tsx [hunk 4]
```diff
@@ -183,36 +303,84 @@ export default function SessionPage() {
...
+        applyMuxSessionPatches(restSessions, pendingMuxSessionsRef.current ?? []) ?? restSessions;
+      cachedSidebarSessions = nextSessions;
+      setSidebarError(false);
+      setSidebarSessions((current) => (
+        areSidebarSessionsEqual(current, nextSessions) ? current : nextSessions
+      ));
+    } catch (err) {
+      console.error("Failed to fetch sidebar sessions:", err);
+      setSidebarError(true);
+      setSidebarSessions((current) => (current === null ? [] : current));
```
Comment (illegalcall) [inline packages/web/src/app/sessions/[id]/page.tsx:335]: On fetch failure, this keeps rendering the previously cached sessions while `sidebarError` is true. `ProjectSidebar` only renders the error branch when `visibleSessions.length === 0` (ProjectSidebar.tsx:257), so when the cache has stale sessions the user sees them with no indication the fetch failed. Should the cache be cleared on error, or the error surfaced alongside the list?

File: packages/web/src/app/sessions/[id]/page.tsx [hunk 5]
```diff
@@ -183,36 +303,84 @@ export default function SessionPage() {
...
     }
   }, []);
 
+  useEffect(() => {
+    if (!mux?.sessions) return;
```
Comment (illegalcall) [inline packages/web/src/app/sessions/[id]/page.tsx:340]: This effect doesn't check `mux.status`. After a WebSocket disconnect `mux.sessions` retains the last snapshot, and `pendingMuxSessionsRef` continues to overlay it onto every REST refresh (line 319–320). If the server changes a session's status/activity/lastActivityAt while the socket is down, REST will fetch the new value and it gets silently overwritten with the stale mux snapshot. Should this gate on `mux.status === "connected"` and clear `pendingMuxSessionsRef` on disconnect?

File: packages/web/src/app/sessions/[id]/page.tsx [hunk 6]
```diff
@@ -183,36 +303,84 @@ export default function SessionPage() {
...
       }
       setZoneCounts(counts);
-    } catch {
-      // non-critical - status strip just won't show
+    } catch (err) {
+      console.error("Failed to fetch project sessions for", projectId, err);
     }
   }, []);
 
   const fetchSidebarSessions = useCallback(async () => {
```
Comment (illegalcall) [inline packages/web/src/app/sessions/[id]/page.tsx:312]: No AbortController or request token here. With the 5s poll (line 390), mux-triggered refetches (lines 353, 359), and the manual retry button all able to run concurrently, responses can resolve out of order. The last one wins and blindly overwrites `cachedSidebarSessions` at line 321, so a stale response that predates a newly-created session can replace a fresh response that included it. The sidebar will drop the new session until the next poll resolves. Should we track a request token and ignore responses whose token isn't current?

**Review by illegalcall — CHANGES_REQUESTED** [review]

File: packages/web/src/app/sessions/[id]/page.tsx (hunk 6 above)
Comment (ashish921998) [reply packages/web/src/app/sessions/[id]/page.tsx:312]: Fixed in 5afa863f — `fetchSidebarSessions` now increments a `sidebarFetchIdRef` counter per invocation and discards any response (success or error) whose token is no longer the latest, so concurrent calls from the 5s poll, mux-triggered refetches, and the retry button can't overwrite a newer response.

File: packages/web/src/app/sessions/[id]/page.tsx (hunk 5 above)
Comment (ashish921998) [reply packages/web/src/app/sessions/[id]/page.tsx:340]: Fixed in 5afa863f — the mux effect now gates on `mux.status === "connected"` and clears `pendingMuxSessionsRef` when the socket is not live, so REST refreshes after a disconnect no longer get overlaid with a stale mux snapshot. Added `mux?.status` to the dep array so the clear runs on transition.

File: packages/web/src/app/sessions/[id]/page.tsx (hunk 4 above)
Comment (ashish921998) [reply packages/web/src/app/sessions/[id]/page.tsx:335]: Fixed in 5afa863f — went with "surface alongside" to avoid flashing an error over still-useful cached data. `ProjectSidebar` now renders a compact banner ("Failed to refresh · showing cached sessions" + Retry) above the tree whenever `error && sessions.length > 0`, so the fetch failure is visible even though the cached list is still shown. The empty-state error branch at line 257 is unchanged.

File: packages/web/src/app/sessions/[id]/page.tsx [hunk 7]
```diff
@@ -183,36 +304,98 @@ export default function SessionPage() {
...
+      setSidebarSessions(next);
+    }
+
+    if (mux.sessions.length === 0 || !cachedSidebarSessions) {
+      return;
+    }
+
+    const cachedIds = new Set(cachedSidebarSessions.map((sidebarSession) => sidebarSession.id));
+    const muxIds = new Set(mux.sessions.map((muxSession) => muxSession.id));
+    if (cachedIds.size !== muxIds.size) {
```
Comment (illegalcall) [inline packages/web/src/app/sessions/[id]/page.tsx:367]: Can you double-check this path? This mismatch check triggers a full `/api/sessions` refetch on every mux snapshot where cached worker IDs and mux session IDs differ. If the mux `sessions` topic publishes all sessions (workers + orchestrators) but the sidebar only caches workers (via `filterWorkerSessions`), the two sets will never match and this will fetch on every snapshot indefinitely. If the server-side topic already filters to workers, disregard — but worth confirming the mux snapshot source applies the same filter as `/api/sessions`.

**Review by illegalcall — APPROVED** [review]

File: packages/web/src/app/sessions/[id]/page.tsx (hunk 7 above)
Comment (ashish921998) [reply packages/web/src/app/sessions/[id]/page.tsx:367]: Good catch — double-checked and the two sources are symmetric:

- **`/api/sessions`** (sidebar fetch, `packages/web/src/app/api/sessions/route.ts`): runs `filterProjectSessions`, then the inline orchestrator filter at lines 64–71 (`!isOrchestratorSession(...)`) → workers only.
- **`/api/sessions/patches`** (mux snapshot source, `packages/web/server/mux-websocket.ts` fetches it at line 105): calls `filterWorkerSessions` in `packages/web/src/app/api/sessions/patches/route.ts:20` → same workers-only set.

`filterWorkerSessions` is just `isOrchestratorSession`-filter + `filterProjectSessions` (see `packages/web/src/lib/project-utils.ts:42`), so both endpoints apply the same worker filter. The cached sidebar set and the mux snapshot set contain the same session IDs, so the mismatch check only triggers on genuine session add/remove — not on every snapshot.
